File: kafka-streams-java/9-processor-api/src/main/java/com/github/simplesteph/udemy/kafka/streams/StatusFilterTransformer.java
package com.github.simplesteph.udemy.kafka.streams;

import com.github.simplesteph.Victory;
import org.apache.kafka.streams.kstream.ValueTransformer;
import org.apache.kafka.streams.processor.ProcessorContext;

import java.nio.charset.StandardCharsets;
import java.util.Optional;

import static com.github.simplesteph.udemy.kafka.streams.LowLevelProcessorApp.ERROR;

public class StatusFilterTransformer implements ValueTransformer<Victory, Victory> {

    private ProcessorContext context;

    @Override
    public void init(ProcessorContext context) {
        this.context = context;
    }

    @Override
    public Victory transform(Victory value) {
        String status = Optional
                .ofNullable(context.headers().lastHeader("status"))
                .map(header -> new String(header.value(), StandardCharsets.UTF_8))
                .orElse("missing-header");

        return status.equals("invalid") ? ERROR : value;
    }

    @Override
    public void close() {

    }
}

File: kafka-streams-scala/9-processor-api-scala/src/main/scala/com/github/simplesteph/udemy/kafka/streams/StatusFilterTransformer.scala
package com.github.simplesteph.udemy.kafka.streams

import com.github.simplesteph.Victory
import org.apache.kafka.streams.kstream.ValueTransformer
import org.apache.kafka.streams.processor.ProcessorContext
import java.nio.charset.StandardCharsets

import com.github.simplesteph.udemy.kafka.streams.LowLevelProcessorApp.VictoryError


class StatusFilterTransformer extends ValueTransformer[Victory, Victory] {

  private var context: ProcessorContext = _

  override def init(context: ProcessorContext): Unit = {
    this.context = context
  }

  override def transform(value: Victory): Victory = {
    val status = Option(context.headers().lastHeader("status"))
      .map(header => new String(header.value(), StandardCharsets.UTF_8))
      .getOrElse("missing-header")

    if (status == "invalid") VictoryError else value
  }

  override def close(): Unit = {}
}
